Web Inspector: Support HTTP throttling or some Network Conditioner for the inspected web page

dcrousso opened the pull request:
#### a8912b8763857e0ab81f5e16f53801f6aaa81ed4
<pre>
Web Inspector: Support HTTP throttling or some Network Conditioner for the inspected web page
<a href="https://bugs.webkit.org/show_bug.cgi?id=147909">https://bugs.webkit.org/show_bug.cgi?id=147909</a>
&lt;rdar://problem/24744411&gt;

Reviewed by Patrick Angle.

Add experimental support for limiting the total bytes per second of each related network task.

* Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
(WI.NetworkManager.displayNameForEmulatedCondition): Added.
(WI.NetworkManager.bytesPerSecondLimitForEmulatedCondition): Added.
(WI.NetworkManager.prototype.initializeTarget):
(WI.NetworkManager.prototype.get emulatedCondition): Added.
(WI.NetworkManager.prototype.set emulatedCondition): Added.
(WI.NetworkManager.prototype._applyEmulatedCondition): Added.
Create a list of known conditions that convert to known QoS values.
- 3G (780kbps)
- DSL (2mbps)
- Edge (240kbps)
- LTE (50mbps)
- Wi-Fi (40mbps)
- Wi-Fi 802.11ac (250mbps)

* Source/WebInspectorUI/UserInterface/Views/NetworkTabContentView.js:
(WI.NetworkTabContentView):
(WI.NetworkTabContentView.tabInfo):
(WI.NetworkTabContentView.prototype._handleEmulatedConditionChanged): Added.
Change the icon of the Network Tab when throttling is active so it&apos;s less likely to be forgotten.

* Source/WebInspectorUI/UserInterface/Views/NetworkTableContentView.js:
(WI.NetworkTableContentView):
(WI.NetworkTableContentView.prototype.get navigationItems):
(WI.NetworkTableContentView.prototype.closed):
(WI.NetworkTableContentView.prototype._handleNetworkConditionSelectionChanged): Added.
* Source/WebInspectorUI/UserInterface/Views/NetworkTableContentView.css:
(.network-condition-scope-bar.default-item-selected:not(:hover)): added.
(.network-condition-scope-bar.default-item-selected:hover): added.
Add a `WI.ScopeBar` for all the known conditions above.

* Source/JavaScriptCore/inspector/protocol/Network.json:
* Source/WebCore/inspector/agents/InspectorNetworkAgent.h:
* Source/WebCore/inspector/agents/InspectorNetworkAgent.cpp:
(WebCore::InspectorNetworkAgent::disable):
(WebCore::InspectorNetworkAgent::setEmulatedConditions): Added.
* Source/WebCore/inspector/agents/page/PageNetworkAgent.h:
* Source/WebCore/inspector/agents/page/PageNetworkAgent.cpp:
(WebCore::PageNetworkAgent::PageNetworkAgent):
(WebCore::PageNetworkAgent::setEmulatedConditionsInternal): Added.
* Source/WebCore/inspector/agents/page/WorkerNetworkAgent.h:
* Source/WebCore/inspector/agents/page/WorkerNetworkAgent.cpp:
(WebCore::WorkerNetworkAgent::setEmulatedConditionsInternal): Added.
* Source/WebCore/inspector/InspectorClient.h:
(WebCore::InspectorClient::setEmulatedConditions): Added.
* Source/WebKit/WebProcess/Inspector/WebInspectorClient.h:
* Source/WebKit/WebProcess/Inspector/WebInspectorClient.cpp:
(WebKit::WebInspectorClient::setEmulatedConditions): Added.
* Source/WebKit/WebProcess/Inspector/WebInspector.h:
* Source/WebKit/WebProcess/Inspector/WebInspector.cpp:
(WebKit::WebInspector::setEmulatedConditions): Added.
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.messages.in:
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.h:
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.cpp:
(WebKit::WebInspectorUIProxy::setEmulatedConditions): Added.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::setEmulatedConditions): Added.
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::setEmulatedConditions): Added.
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::setEmulatedConditions): Added.
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::registerNetworkDataTask):
(WebKit::NetworkSession::setEmulatedConditions): Added.
* Source/WebKit/NetworkProcess/NetworkDataTask.h:
(WebKit::NetworkDataTask::setEmulatedConditions): Added.
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
(WebKit::NetworkDataTaskCocoa::~NetworkDataTaskCocoa):
(WebKit::NetworkDataTaskCocoa::setEmulatedConditions): Added.
Pass along the QoS values to all related `NSURLSessionTask` whenever setting the emulated condition.

* Source/WebCore/inspector/InspectorController.cpp:
(WebCore::InspectorController::createLazyAgents):
Pass along the `InspectorClient` to the `PageNetworkAgent` too.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
Create a new `ENABLE_INSPECTOR_NETWORK_THROTTLING` since this is currently only available on cocoa.

* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype._createExperimentalSettingsView):
Add an experimental setting for showing this in Web Inspector.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* LayoutTests/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/http/tests/inspector/network/setEmulatedConditions.html: Added.
* LayoutTests/http/tests/inspector/network/setEmulatedConditions-expected.txt: Added.

Canonical link: <a href="https://commits.webkit.org/252589@main">https://commits.webkit.org/252589@main</a>
</pre>

JosephPecoraro commented on Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
```diff
@@ -193,7 +253,7 @@ WI.NetworkManager = class NetworkManager extends WI.Object
             target.NetworkAgent.enable();
 
-            // COMPATIBILITY (iOS 10.3): Network.setDisableResourceCaching did not exist.
+            // COMPATIBILITY (iOS 10.3): Network.setResourceCachingDisabled did not exist.
```
😄

dcrousso replied:
(☞ﾟヮﾟ)☞

couldn't resist it lol

JosephPecoraro commented on Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
```diff
@@ -422,15 +422,19 @@ static inline bool computeIsAlwaysOnLoggingAllowed(NetworkSession& session)
-    if (!m_task || !m_sessionWrapper)
-        return;
+    if (m_task && m_sessionWrapper) {
+        RELEASE_ASSERT(m_sessionWrapper->dataTaskMap.get([m_task taskIdentifier]) == this);
```
Could pull `[m_task taskIdentifier]` out into its own local to avoid a potential double msg send.

JosephPecoraro commented on Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
```diff
@@ -763,6 +767,16 @@ static inline bool computeIsAlwaysOnLoggingAllowed(NetworkSession& session)
+
+void NetworkDataTaskCocoa::setEmulatedConditions(const std::optional<int64_t>& bytesPerSecondLimit)
+{
+    if (bytesPerSecondLimit)
```
I wonder if somebody should assert this is positive... I never liked `int64_t`.

patrickangle replied:
Is there a reason we aren't using uint64_t instead? Seems like we are wasting half of `int64_t`'s potential since it can't be lower than 0.

dcrousso replied:
CFNetwork uses `int64_t` for their SPI

patrickangle commented on Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
```diff
@@ -1607,6 +1703,16 @@ WI.NetworkManager = class NetworkManager extends WI.Object
     }
 };
 
+WI.NetworkManager.EmulatedCondition = {
```
Would it make more sense for these to be objects that contain their display name, bpsLimit, and other potential future data? Having two static methods to extract that info given one of these values already seems like a lot, and I fear that would only get worse if/when we are able to implement latency, for example.

dcrousso replied:
I dont believe we've ever done a `WI.UIString` on load.  It should theoretically be possible, but I'll have to check and make sure.

Most of this was just to avoid having to do all that work on load, but it's probably a drop in the bucket.  I'll rework.

JosephPecoraro commented on Source/WebCore/inspector/agents/InspectorNetworkAgent.cpp:
```diff
@@ -908,6 +908,10 @@ Protocol::ErrorStringOr<void> InspectorNetworkAgent::disable()
     setResourceCachingDisabled(false);
 
+#if ENABLE(INSPECTOR_NETWORK_THROTTLING)
+    setEmulatedConditions(0);
```
Probably `std::nullopt` instead of `0`?

dcrousso replied:
`std::nullopt` is more of a "leave this value as it is"

the idea is that later on if/when we add packet loss, latency, etc. then we have a way of saying "only adjust this" instead of having to provide everything

in this case we explicitly want to undo any previously applied throttling, so we reset back to the initial value of `0`

patrickangle replied:
I agree with @JosephPecoraro, since for different network conditions `0` can mean something different. `0` bps being a value that makes no sense in practice, I understand why `0` is a magic value here, but if we were to add something like `latencyMilliseconds`, 0 would be a perfectly reasonable value. Sure, 0 is also the default behavior where no latency is added, but it would, IMO, be clearer if std:nullopt returned the value to defaults for every property. What you've done also works, it just feels like std:nullopt would be a clearer "use the default" value.

Optimizing to, in the future, avoid sending an additional integer or two over the protocol in response to a user's choice seems premature to me.

dcrousso replied:
Yeah i suppose we could just make it so that anyone calling `Network.setEmulatedConditions` has to provide values for everything they want to affect up front instead of having the backend maintain this sorta implicit state.

JosephPecoraro commented on Source/WebCore/inspector/agents/page/PageNetworkAgent.cpp:
```diff
@@ -40,14 +41,29 @@ namespace WebCore {
+        return { };
+    }
+
+    return InspectorNetworkAgent::setEmulatedConditions(WTFMove(bytesPerSecondLimit));
```
Err, does this end up returning "unsupported"?

dcrousso replied:
yes that's correct (and desired)

if we don't have a `m_client` then we kinda are in an unsupported state :)

JosephPecoraro commented on Source/WebCore/inspector/agents/page/PageNetworkAgent.cpp:
```diff
@@ -40,14 +41,29 @@ namespace WebCore {
+
+Protocol::ErrorStringOr<void> PageNetworkAgent::setEmulatedConditions(std::optional<int>&& bytesPerSecondLimit)
+{
+    if (m_client) {
```
Here might be a good place to validate the value is positive.

```cpp
    if (bytesPerSecondLimit && *bytesPerSecondLimit < 0)
        return makeUnexpected("Unexpected negative bytesPerSecond"_s);
```

JosephPecoraro replied:
Or perhaps putting that in the super class and doing that operation first before messaging the client.

dcrousso replied:
does `-[NSURLSessionTask _setBytesPerSecondLimit:]` disallow negatives?  i can add it if so, but wasn't sure (and couldn't really grok the code in CFNetwork 😅)

patrickangle reviewed (approved):
Awesome - I think some folks are going to be very excited. Just a few comments/questions/suggestions.

dcrousso committed:
Web Inspector: Support HTTP throttling or some Network Conditioner for the inspected web page
https://bugs.webkit.org/show_bug.cgi?id=147909
<rdar://problem/24744411>

Reviewed by Patrick Angle.

Add experimental support for limiting the total bytes per second of each related network task.

* Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
(WI.NetworkManager.displayNameForEmulatedCondition): Added.
(WI.NetworkManager.bytesPerSecondLimitForEmulatedCondition): Added.
(WI.NetworkManager.prototype.initializeTarget):
(WI.NetworkManager.prototype.get emulatedCondition): Added.
(WI.NetworkManager.prototype.set emulatedCondition): Added.
(WI.NetworkManager.prototype._applyEmulatedCondition): Added.
Create a list of known conditions that convert to known QoS values.
- 3G (780kbps)
- DSL (2mbps)
- Edge (240kbps)
- LTE (50mbps)
- Wi-Fi (40mbps)
- Wi-Fi 802.11ac (250mbps)

* Source/WebInspectorUI/UserInterface/Views/NetworkTabContentView.js:
(WI.NetworkTabContentView):
(WI.NetworkTabContentView.tabInfo):
(WI.NetworkTabContentView.prototype._handleEmulatedConditionChanged): Added.
Change the icon of the Network Tab when throttling is active so it's less likely to be forgotten.

* Source/WebInspectorUI/UserInterface/Views/NetworkTableContentView.js:
(WI.NetworkTableContentView):
(WI.NetworkTableContentView.prototype.get navigationItems):
(WI.NetworkTableContentView.prototype.closed):
(WI.NetworkTableContentView.prototype._handleNetworkConditionSelectionChanged): Added.
* Source/WebInspectorUI/UserInterface/Views/NetworkTableContentView.css:
(.network-condition-scope-bar.default-item-selected:not(:hover)): added.
(.network-condition-scope-bar.default-item-selected:hover): added.
Add a `WI.ScopeBar` for all the known conditions above.

* Source/JavaScriptCore/inspector/protocol/Network.json:
* Source/WebCore/inspector/agents/InspectorNetworkAgent.h:
* Source/WebCore/inspector/agents/InspectorNetworkAgent.cpp:
(WebCore::InspectorNetworkAgent::disable):
(WebCore::InspectorNetworkAgent::setEmulatedConditions): Added.
* Source/WebCore/inspector/agents/page/PageNetworkAgent.h:
* Source/WebCore/inspector/agents/page/PageNetworkAgent.cpp:
(WebCore::PageNetworkAgent::PageNetworkAgent):
(WebCore::PageNetworkAgent::setEmulatedConditionsInternal): Added.
* Source/WebCore/inspector/agents/page/WorkerNetworkAgent.h:
* Source/WebCore/inspector/agents/page/WorkerNetworkAgent.cpp:
(WebCore::WorkerNetworkAgent::setEmulatedConditionsInternal): Added.
* Source/WebCore/inspector/InspectorClient.h:
(WebCore::InspectorClient::setEmulatedConditions): Added.
* Source/WebKit/WebProcess/Inspector/WebInspectorClient.h:
* Source/WebKit/WebProcess/Inspector/WebInspectorClient.cpp:
(WebKit::WebInspectorClient::setEmulatedConditions): Added.
* Source/WebKit/WebProcess/Inspector/WebInspector.h:
* Source/WebKit/WebProcess/Inspector/WebInspector.cpp:
(WebKit::WebInspector::setEmulatedConditions): Added.
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.messages.in:
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.h:
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.cpp:
(WebKit::WebInspectorUIProxy::setEmulatedConditions): Added.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::setEmulatedConditions): Added.
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::setEmulatedConditions): Added.
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::setEmulatedConditions): Added.
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::registerNetworkDataTask):
(WebKit::NetworkSession::setEmulatedConditions): Added.
* Source/WebKit/NetworkProcess/NetworkDataTask.h:
(WebKit::NetworkDataTask::setEmulatedConditions): Added.
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
(WebKit::NetworkDataTaskCocoa::~NetworkDataTaskCocoa):
(WebKit::NetworkDataTaskCocoa::setEmulatedConditions): Added.
Pass along the QoS values to all related `NSURLSessionTask` whenever setting the emulated condition.

* Source/WebCore/inspector/InspectorController.cpp:
(WebCore::InspectorController::createLazyAgents):
Pass along the `InspectorClient` to the `PageNetworkAgent` too.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
Create a new `ENABLE_INSPECTOR_NETWORK_THROTTLING` since this is currently only available on cocoa.

* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype._createExperimentalSettingsView):
Add an experimental setting for showing this in Web Inspector.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* LayoutTests/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/http/tests/inspector/network/setEmulatedConditions.html: Added.
* LayoutTests/http/tests/inspector/network/setEmulatedConditions-expected.txt: Added.

Canonical link: https://commits.webkit.org/252589@main
